pkg/apis/sources/v1alpha1/gcp_pubsub_types.go

```diff
@@ -0,0 +1,163 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/knative/pkg/apis/duck"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Important: Run "make generate" to regenerate code after modifying this file
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// GcpPubSubSource is the Schema for the gcppubsubsources API.
+// +k8s:openapi-gen=true
+// +kubebuilder:categories=all,knative,eventing,sources
+type GcpPubSubSource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   GcpPubSubSourceSpec   `json:"spec,omitempty"`
+	Status GcpPubSubSourceStatus `json:"status,omitempty"`
+}
+
+// Check that GcpPubSubSource can be validated and can be defaulted.
+var _ runtime.Object = (*GcpPubSubSource)(nil)
+
+// Check that GcpPubSubSource implements the Conditions duck type.
+var _ = duck.VerifyType(&GcpPubSubSource{}, &duckv1alpha1.Conditions{})
+
+// GcpPubSubSourceSpec defines the desired state of the GcpPubSubSource.
+type GcpPubSubSourceSpec struct {
+	// GcpCredsSecret is the credential to use to poll the GCP PubSub Subscription. It is not used
+	// to create or delete the Subscription, only to poll it. The value of the secret entry must be
+	// a service account key in the JSON format (see
+	// https://cloud.google.com/iam/docs/creating-managing-service-account-keys).
+	GcpCredsSecret corev1.SecretKeySelector `json:"gcpCredsSecret,omitempty"`
+
+	// GoogleCloudProject is the ID of the Google Cloud Project that the PubSub Topic exists in.
+	GoogleCloudProject string `json:"googleCloudProject,omitempty"`
+
+	// Topic is the ID of the GCP PubSub Topic to Subscribe to. It must be in the form of the
+	// unique identifier within the project, not the entire name. E.g. it must be 'laconia', not
+	// 'projects/my-gcp-project/topics/laconia'.
+	Topic string `json:"topic,omitempty"`
+
+	// Sink is a reference to an object that will resolve to a domain name to use as the sink.
+	// +optional
+	Sink *corev1.ObjectReference `json:"sink,omitempty"`
+}
+
+const (
+	// GcpPubSubConditionReady has status True when the GcpPubSubSource is ready to send events.
+	GcpPubSubConditionReady = duckv1alpha1.ConditionReady
+
+	// GcpPubSubConditionSinkProvided has status True when the GcpPubSubSource has been configured with a sink target.
+	GcpPubSubConditionSinkProvided duckv1alpha1.ConditionType = "SinkProvided"
```

Comment:
I wonder if we should make a followup to consolidate this for sources.

Reply:
At the very least a few of the conditions are in common for all Sources, so I think it would be nice if we consolidate them.